evalml/pipelines/components/transformers/preprocessing/stl_decomposer.py

```diff
@@ -158,43 +173,81 @@ def fit(self, X: pd.DataFrame, y: pd.Series = None) -> STLDecomposer:
             ValueError: If y is None.
             ValueError: If target data doesn't have DatetimeIndex AND no Datetime features in features data
         """
-        self.original_index = y.index if y is not None else None
-        X, y = self._check_target(X, y)
-        self._map_dt_to_integer(self.original_index, y.index)
+        from evalml.pipelines.utils import unstack_multiseries
 
         # Warn for poor decomposition use with higher seasonal smoothers
         if self.seasonal_smoother > 14:
             self.logger.warning(
                 f"STLDecomposer may perform poorly on data with a high seasonal smoother ({self.seasonal_smoother}).",
             )
 
+        # If y is a stacked pd.Series, unstack it
+        if self.series_id is not None and isinstance(y, pd.Series):
+            X, y = unstack_multiseries(X, y, self.series_id, self.time_index, y.name)
+
+        if isinstance(y, pd.Series):
+            y = y.to_frame()
+
+        self.original_index = y.index if y is not None else None
+
+        X, y = self._check_target(X, y)
+
+        self._map_dt_to_integer(self.original_index, y.index)
+
         # Save the frequency of the fitted series for checking against transform data.
         self.frequency = y.index.freqstr or pd.infer_freq(y.index)
-
-        # Determine the period of the seasonal component
-        if self.period is None:
-            self.set_period(X, y)
-
-        stl = STL(y, seasonal=self.seasonal_smoother, period=self.period)
-        res = stl.fit()
-        self.seasonal = res.seasonal
-        self.period = stl.period
-        dist = len(y) % self.period
-        self.seasonality = (
-            self.seasonal[-(dist + self.period) : -dist]
-            if dist > 0
-            else self.seasonal[-self.period :]
-        )
-        self.trend = res.trend
-        self.residual = res.resid
+        # Iterate through each id group
+        self.seasonals = {}
+        self.seasonalities = {}
+        self.trends = {}
+        self.residuals = {}
+        if self.periods is None:
+            self.periods = {}
+
+        for id in y.columns:
+            series_y = y[id]
+
+            # Determine the period of the seasonal component
+            if id not in self.periods:
+                period = self.determine_periodicity(
+                    X,
+                    series_y,
+                    acf_threshold=0.01,
+                    rel_max_order=5,
+                )
+                if self.period is None and len(y.columns) == 1:
+                    self.period = period
+                    self.update_parameters({"period": self.period})
+                elif self.period is not None and len(y.columns) == 1:
+                    period = self.period
```

Comment:
Since everything is accessing the period through `self.periods` I'm thinking we might not even need to update `self.period` at all, instead just use it to set `period` if its given by a user